config/configsource/configsource.go

```diff
@@ -0,0 +1,267 @@
+	allKeys := srcCfg.AllKeys()
+	sort.Slice(allKeys, deepestConfigSourcesFirst(allKeys))
+
+	for _, k := range allKeys {
+		dstKey, cfgSrcName, paramsKey := extractCfgSrcInvocation(k)
+		if cfgSrcName == "" {
+			// Nothing to apply take the key and value as it is.
+			dstCfg.Set(k, srcCfg.Get(k))
+			continue
+		}
+
+		if _, ok := appliedTags[dstKey]; ok {
+			// Already applied this key.
+			continue
+		}
+
+		appliedTags[dstKey] = struct{}{}
+		cfgSrc, ok := cfgSources[cfgSrcName]
+		if !ok {
+			return false, &cfgSrcError{
+				msg:  fmt.Sprintf("config source %s not found", cfgSrcName),
+				code: errCfgSrcNotFound,
+			}
+		}
+
+		applyParams := srcCfg.Get(paramsKey)
+		actualCfg, err := cfgSrc.Apply(ctx, applyParams)
+		if err != nil {
+			return false, &cfgSrcError{
+				msg:  fmt.Sprintf("error applying config source %s: %v", cfgSrcName, err),
+				code: errCfgSrcApply,
+			}
+		}
+
+		// The injection may require further expansion, assume that we are not done yet.
+		done = false
+		if dstKey != "" {
+			dstCfg.Set(dstKey, actualCfg)
+			continue
+		}
+
+		// This is at the root level, have to inject the top keys one by one.
+		rootMap, ok := actualCfg.(map[string]interface{})
+		if !ok {
+			return false, &cfgSrcError{
+				msg:  "only a map can be injected at the root level",
+				code: errOnlyMapAtRootLevel,
+			}
+		}
+		for k, v := range rootMap {
+			dstCfg.Set(k, v)
+		}
+	}
+
+	return done, nil
+}
+
+func deepestConfigSourcesFirst(keys []string) func(int, int) bool {
+	return func(i, j int) bool {
+		iLastSrcIdx := strings.LastIndex(keys[i], ConfigSourcePrefix)
+		jLastSrcIdx := strings.LastIndex(keys[j], ConfigSourcePrefix)
+		if iLastSrcIdx != jLastSrcIdx {
+			// At least one of the keys has a config source.
+			return iLastSrcIdx > jLastSrcIdx
+		}
+
+		// It can be one of the following cases:
+		//  1. None has a config source invocation
+		//  2. Both have at the same level, ie.: they are siblings in the config.
+		// In any case it doesn't matter which one is considered "smaller".
+		return true
+	}
+}
+
+func extractCfgSrcInvocation(k string) (dstKey, cfgSrcName, paramsKey string) {
+	firstPrefixIdx := strings.Index(k, ConfigSourcePrefix)
+	if firstPrefixIdx == -1 {
+		// No config source to be applied.
+		return
+	}
+
+	// Check for a deeper one.
+	tagPrefixIdx := strings.LastIndex(k, ConfigSourcePrefix)
+	dstKey = strings.TrimSuffix(k[:tagPrefixIdx], config.ViperDelimiter)
+
+	cfgSrcFromStart := k[tagPrefixIdx+len(ConfigSourcePrefix):]
+	prefixToTagEndLen := strings.Index(cfgSrcFromStart, config.ViperDelimiter)
+	if prefixToTagEndLen > -1 {
+		cfgSrcName = cfgSrcFromStart[:prefixToTagEndLen]
+	} else {
+		cfgSrcName = cfgSrcFromStart
+	}
+
+	paramsKey = k[:tagPrefixIdx+len(cfgSrcName)+len(ConfigSourcePrefix)]
+
+	return
+}
+
+// Copied from config package to avoid exposing as public API.
+// TODO: Add local tests covering the code below.
```

Comment:
Consider to move it to `/config/internal` so even if it is public is not exposed.